foundry/glm/glm.py

```diff
@@ -681,30 +703,34 @@ def estimate_laplace_coefs(self, X: ModelMatrix, y: ModelMatrix, sample_weight:
 
         # create mvnorm for laplace approx:
         with torch.no_grad():
-            fail_msg = ''
+
+            self.converged_ = False
+
             try:
-                cov = torch.inverse(hess)
-            except RuntimeError as e:
-                fail_msg = str(e)
+                # cov = torch.inverse(hess)  # Hello floating point errors my old friend.
+                cholesky_hess = torch.linalg.cholesky(hess)  # This is way more numerically stable
+                inv_cholesky_hess = torch.linalg.inv(cholesky_hess)
+                cov = inv_cholesky_hess.T @ inv_cholesky_hess
 
-            if not fail_msg:
-                try:
-                    self._coef_mvnorm_ = torch.distributions.MultivariateNormal(
-                        means, covariance_matrix=cov, validate_args=True
-                    )
-                    self.converged_ = True
-                except ValueError as e:
-                    if 'constraint PositiveDefinite' not in str(e):
-                        raise e
-                    fail_msg = str(e)
-
-            if fail_msg:
-                warn(f"Model failed to converge; ``laplace_params`` cannot be estimated."
-                     f"\nhess.diag():\n{hess.diag()}"
-                     f"\nfail msg:\n{fail_msg}")
-                fake_cov = torch.eye(hess.shape[0]) * 1e-10
+                # if `hess` has one eigenvalue of zero, then the matrix will be non-invertible. This usually means one
+                # of the features is constant, or is completely colinear with another feature.
+
+                # if `hess` is poorly conditioned, but is positive definite (PD), it may not remain PD after inversion.
+                # Using the Cholesky decomposition helps, but can still cause issues
+
+                # if `hess` is negative definite (ND), the model hasn't converged. Consider looking at the stopping criterion.
+
+                self._coef_mvnorm_ = torch.distributions.MultivariateNormal(
+                    means, covariance_matrix=cov, validate_args=True
+                )
+
+                self.converged_ = True
+
+            except RuntimeError as e:
+                warn(f"RunTimeError(\"{(str(e))}\")")
+                warn("Second order estimation of parameter distribution failed. Constructing approximate covariance.")
+                fake_cov = torch.diag(torch.diag(hess).pow(-1).clip(min=1E-5))
```

Comment:
I have not so great feelings about constructing a fake-but-realistic covariance matrix. What would it be used for? Could this result in confusion?

Reply:
In the same way that a hessian allows for second-order gradient descent, this fake-but-realistic is used for first-order gradient descent. 
Using the same logic, sampling from this is like a first-order estimation of the variation of model coefs.